core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java

```diff
@@ -768,6 +775,28 @@ public PlanOptimizers(
                             // Must run before AddExchanges and after ReplicateSemiJoinInDelete
                             // to avoid temporarily having an invalid plan
                             new DetermineSemiJoinDistributionType(costComparator, taskCountEstimator)))));
+
+            // Must run before DetermineTableScanNodePartitioning because PushJoinIntoTableScan creates new table handles
+            // without useConnectorNodePartitioning set which will cause planner to fail
+            builder.add(new IterativeOptimizer(
+                    ruleStats,
+                    statsCalculator,
+                    estimatedExchangesCostCalculator,
+                    pushIntoTableScanRules));
+
+            // DetermineTableScanNodePartitioning needs to ensure all table handles have proper partitioning determined
+            builder.add(new IterativeOptimizer(
+                    ruleStats,
+                    statsCalculator,
+                    costCalculator,
+                    ImmutableSet.of(
+                            // Must run before AddExchanges
+                            new PushDeleteIntoConnector(metadata),
+                            new DetermineTableScanNodePartitioning(metadata, nodePartitioningManager, taskCountEstimator),
+                            // Must run after join reordering because join reordering creates
+                            // new join nodes without JoinNode.maySkipOutputDuplicates flag set
```

Comment:
reordering -> pushdown ?

(twice)

Reply:
The comment is correct (but indeed confusing - I'll reword). What I meant that this rule needs to `run after join reordering because it (join reordering) creates new join nodes ....`

```
                            // Must run after join reordering because it creates
                            // new join nodes without JoinNode.maySkipOutputDuplicates flag set
```

WDYT?